### PR TITLE
feat: gitlab authentication integration - GSoC'22 Vedant Jain

### LINKED ISSUE #3157: Gitlab integration
## Description
Implement sign in with Gitlab

### PR DESCRIPTION
**Pull request by vedant-jain03**
fixes #3157 

#### Describe the changes you have made in this PR -

- Added `omniauth-gitlab` gem into `Gemfile`.
- Configure files
- Added SVG icon
- Added views code
- Added env variables

### CODE REVIEW

**Comment by tachyons** [comment]
Good job, Could you also make this under the feature flag using flipper ?

**Comment by vedant-jain03** [comment]
> Good job, Could you also make this under the feature flag using flipper ?

Thanks, Sure will make it.

**Comment by vedant-jain03** [comment]
@tachyons 
the gitlab API is returning `404 error not found`. Have I done something wrong?

**Comment by tachyons** [comment]
Likely, have you configured the env vars correctly ?

**Comment by vedant-jain03** [comment]
> Likely, have you configured the env vars correctly ?

Yes, I followed the following steps:

### Creating App
- I create a new application from [here](https://gitlab.com/-/profile/applications).
- I added the name.
- Redirect URI
```
http://localhost:3000/import/gitlab/callback
http://localhost:3000/users/auth/gitlab/callback
```
- Enabled scope : read_user.
- After creating the application, I copied the Application ID, and Secret and pasted them into `.env`.

**Comment by tachyons** [comment]
> http://localhost:3000/import/gitlab/callback
This is not needed.

Also could you provide more details of the error, error trace, endpoint, response data etc ?

**Comment by vedant-jain03** [comment]
> Member

Using command `curl -I https://gitlab.com/api/v4/oauth/authorize?client_id=<gitlab-id>&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fusers%2Fauth%2Fgitlab%2Fcallback&response_type=code&scope=read_user&state=c0c1817e6f2dd7630a36668c3639c5633cc625856866071a`

Response: 

```
HTTP/2 404 
date: Sun, 05 Jun 2022 06:20:57 GMT
content-type: application/json
cache-control: no-cache
vary: Origin
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-request-id: 01G4S8C5Q0KCXV2N7S0CM0WS2D
x-runtime: 0.016728
strict-transport-security: max-age=31536000
referrer-policy: strict-origin-when-cross-origin
gitlab-lb: fe-19-lb-gprd
gitlab-sv: api-gke-us-east1-c
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=SzKoNvu4B%2BQvD0q%2Ff%2BSe%2BfOWh3KkR787h%2Fg7gfPzAAYUcpVssmxs%2BDby78qu4QRlK%2FTGP%2FecNPCFCvWOL4Awf1v8AhafBWKF0FBWJ6eDB9hvFmAYMMKzzcKZ3mBIu3aPclKaisNwhXw%3D"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}
server: cloudflare
cf-ray: 7166ce2a094a31b2-BOM
```

**Comment by vedant-jain03** [comment]
@abstrekt @tachyons @pavanjoshi914 
This is working fine!

https://user-images.githubusercontent.com/76901313/173349914-d7c3db18-cedb-46c5-99c8-a6900a5fd88e.mp4

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # frozen_string_literal: true
 
 class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  before_action :gitlab, only: %i[show], constraints: lambda {
```
Comment (tachyons) [inline app/controllers/users/omniauth_callbacks_controller.rb:4]: why do we need before action here ?

**Comment by abstrekt** [comment]
> @abstrekt @tachyons @pavanjoshi914 This is working fine!
> 
>  simplescreenrecorder-2022-05-01_09.23.50-2022-06-13_17.33.17-2022-06-13_17.35.41.mp4

can you share a video on how the flow is if we need to sign in to gitlab?

**Comment by vedant-jain03** [comment]
> > @abstrekt @tachyons @pavanjoshi914 This is working fine!
> > simplescreenrecorder-2022-05-01_09.23.50-2022-06-13_17.33.17-2022-06-13_17.35.41.mp4
> 
> can you share a video on how the flow is if we need to sign in to gitlab?


https://user-images.githubusercontent.com/76901313/173353717-e0b1e296-cc8e-4382-8688-9eb92dc20309.mp4

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -26,10 +26,14 @@
   config.omniauth :facebook, ENV['FACEBOOK_CLIENT_ID'], ENV['FACEBOOK_CLIENT_SECRET'],  {
     :info_fields => 'email,name'
   }
-   config.omniauth :github, ENV['GITHUB_CLIENT_ID'], ENV['GITHUB_CLIENT_SECRET'], {
+  config.omniauth :github, ENV['GITHUB_CLIENT_ID'], ENV['GITHUB_CLIENT_SECRET'], {
     :scope => 'read:user'
   }
-
+  if Flipper.enabled?(:gitlab_integration)
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply config/initializers/devise.rb]: Done

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (tachyons) [reply app/controllers/users/omniauth_callbacks_controller.rb:4]: Why do we need to check that for every action action, we need to check that for gitlab action only

Comment (vedant-jain03) [reply app/controllers/users/omniauth_callbacks_controller.rb]: I have added the commit!
Kindly review.

**Comment by tachyons** [comment]
Could you make the codeclimate happy?

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 3]
```diff
@@ -39,6 +39,12 @@ def google_oauth2
   def github
     generic_callback("github")
   end
+  
+  def gitlab
+    if Flipper.enabled?(:gitlab_integration)
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/controllers/users/omniauth_callbacks_controller.rb]: Resolved

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 4]
```diff
@@ -39,6 +39,12 @@
...
```
[the remark this answers is not in the document]
Comment (vedant-jain03) [reply app/controllers/users/omniauth_callbacks_controller.rb]: Resolved

**Comment by tachyons** [comment]
Good job 👏

**Pull request merged**